### PR TITLE
Fix deletion and mass deletion of custom entities

### PR DESCRIPTION
**Pull request by fitn**
Rework fix #125 to also handle mass deletion.
A subscriber on `PRE_REMOVE` event has been created and it checks if the reference data is linked to a product or not.

### CODE REVIEW

File: Event/Subscriber/CheckReferenceDataOnRemoval.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use Akeneo\Component\StorageUtils\Event\RemoveEvent;
+use Akeneo\Component\StorageUtils\StorageEvents;
+use Pim\Bundle\CustomEntityBundle\Remover\NonRemovableEntityException;
+use Pim\Bundle\CustomEntityBundle\Repository\AttributeRepository;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ *
```
Comment (fitn) [inline Event/Subscriber/CheckReferenceDataOnRemoval.php]: Adds phpdoc

File: Event/Subscriber/CheckReferenceDataOnRemoval.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * @author    Romain Monceau <romain@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ */
+class CheckReferenceDataOnRemoval implements EventSubscriberInterface
```
Comment (fitn) [inline Event/Subscriber/CheckReferenceDataOnRemoval.php]: Suffix with Subscriber

File: Resources/config/savers.yml [hunk 3]
```diff
@@ -12,12 +12,11 @@ services:
             - '%pim_reference_data.model.reference_data.interface%'
 
     pim_custom_entity.remover.custom_entity:
-        class: '%pim_custom_entity.remover.custom_entity.class%'
```
Comment (fitn) [inline Resources/config/savers.yml]: Remove class

File: Resources/config/savers.yml [hunk 4]
```diff
@@ -12,12 +12,11 @@ services:
...
+        class: '%pim_catalog.remover.base.class%'
         arguments:
             - '@doctrine.orm.entity_manager'
             - '@event_dispatcher'
             - '%pim_reference_data.model.reference_data.interface%'
-            - '@pim_custom_entity.checker.product_link'
```
Comment (fitn) [inline Resources/config/savers.yml]: Remove checker and checker class

File: Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+use Akeneo\Component\StorageUtils\Event\RemoveEvent;
+use Akeneo\Component\StorageUtils\StorageEvents;
+use Pim\Bundle\CustomEntityBundle\Remover\NonRemovableEntityException;
+use Pim\Bundle\CustomEntityBundle\Repository\AttributeRepository;
+use Pim\Component\Catalog\Query\Filter\Operators;
+use Pim\Component\Catalog\Query\ProductQueryBuilderFactoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Checks if a reference data could be remove or not
```
Comment (mmetayer) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: can be removed

Comment (jmleroux) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: `could be removed`

File: Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * Checks if the reference data is used in a product
+     *
+     * @param RemoveEvent $event
+     */
+    public function checkReferenceDataUsage(RemoveEvent $event)
+    {
+        $referenceData = $event->getSubject();
+
+        $attributes = $this->attributeRepository->findByReferenceData($referenceData);
```
Comment (mmetayer) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: Test if `$referenceData` is an instance of `AbstractCustomEntity` first
Also, maybe `$referenceData` is not a very good name for this variable ;)

File: Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Checks if a reference data could be remove or not
+ *
+ * @author    Romain Monceau <romain@akeneo.com>
+ * @copyright 2018 Akeneo SAS (http://www.akeneo.com)
+ */
+class CheckReferenceDataOnRemovalSubscriber implements EventSubscriberInterface
+{
+    /** @var AttributeRepository */
```
Comment (jmleroux) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: We should maybe create an interface.

File: Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    public function checkReferenceDataUsage(RemoveEvent $event)
+    {
+        $referenceData = $event->getSubject();
+
+        $attributes = $this->attributeRepository->findByReferenceData($referenceData);
+        foreach ($attributes as $attribute) {
+            $pqb = $this->pqbFactory->create();
+            $pqb->addFilter($attribute->getCode(), Operators::IN_LIST, [$referenceData->getCode()]);
+            if (0 !== $pqb->execute()->count()) {
+                throw new NonRemovableEntityException('Reference data "%s" is linked to at least one product');
```
Comment (jmleroux) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: To help in the resolution, or just to make the logs more understandable, we could maybe add the information we already have:
- product count
- attribute code

File: Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Checks if the reference data is used in a product
+     *
+     * @param RemoveEvent $event
+     */
+    public function checkReferenceDataUsage(RemoveEvent $event)
+    {
+        $referenceData = $event->getSubject();
+
```
Comment (jmleroux) [inline Event/Subscriber/CheckReferenceDataOnRemovalSubscriber.php]: Check if we are actually removing a reference data to return early if it's not the case.

**Pull request merged**